File: tracerate/regional.py
import socket
import time


# Linode publishes public speedtest endpoints in each region.
# Stable hostnames, anycast-free, good proxies for "real" geo distance.
REGIONS = [
    ("IN", "Mumbai",     "speedtest.mumbai1.linode.com"),
    ("SG", "Singapore",  "speedtest.singapore.linode.com"),
    ("JP", "Tokyo",      "speedtest.tokyo2.linode.com"),
    ("DE", "Frankfurt",  "speedtest.frankfurt.linode.com"),
    ("UK", "London",     "speedtest.london.linode.com"),
    ("US", "Newark",     "speedtest.newark.linode.com"),
    ("US", "Fremont",    "speedtest.fremont.linode.com"),
]


def _tcp_ping(host: str, port: int = 443, attempts: int = 3, timeout: float = 2.0) -> float:
    """
    TCP-connect latency in ms, returns the minimum of N attempts.
    Min is more representative of true RTT than mean — high samples
    are jitter, the floor is the actual round-trip.
    """

    samples = []
    for _ in range(attempts):
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(timeout)
            start = time.perf_counter()
            sock.connect((host, port))
            samples.append((time.perf_counter() - start) * 1000)
        except (socket.timeout, socket.error):
            pass
        finally:
            try:
                sock.close()
            except Exception:
                pass

    if not samples:
        return 0.0
    return round(min(samples), 2)


def measure_regions() -> list[dict]:
    out = []
    for code, city, host in REGIONS:
        out.append({
            "code": code,
            "city": city,
            "host": host,
            "ms": _tcp_ping(host),
        })
    return out

File: src/regional.rs
use std::time::Instant;
use tokio::net::TcpStream;
use tokio::task;

struct Region {
    code: &'static str,
    city: &'static str,
    host: &'static str,
}

const REGIONS: [Region; 8] = [
    Region {
        code: "IN",
        city: "Mumbai (South Asia)",
        host: "speedtest.mumbai1.linode.com",
    },
    Region {
        code: "SG",
        city: "Singapore (SouthEast Asia)",
        host: "speedtest.singapore.linode.com",
    },
    Region {
        code: "JP",
        city: "Tokyo (East Asia)",
        host: "speedtest.tokyo2.linode.com",
    },
    Region {
        code: "GB",
        city: "London (Europe)",
        host: "speedtest.london.linode.com",
    },
    Region {
        code: "DE",
        city: "Frankfurt (Europe)",
        host: "speedtest.frankfurt.linode.com",
    },
    Region {
        code: "US",
        city: "Newark (US East)",
        host: "speedtest.newark.linode.com",
    },
    Region {
        code: "US",
        city: "Fremont/Seattle (US West)",
        host: "speedtest.fremont.linode.com",
    },
    Region {
        code: "AU",
        city: "Sydney (Australia)",
        host: "speedtest.sydney.linode.com",
    },
];

#[derive(Clone, serde::Serialize)]
pub struct RegionResult {
    pub code: String,
    pub city: String,
    pub host: String,
    pub ms: f64,
}

pub async fn ping_regions() -> Vec<RegionResult> {
    let mut handles = Vec::new();

    for region in &REGIONS {
        let host = region.host;
        let code = region.code;
        let city = region.city;

        let handle = task::spawn(async move {
            let ms = tcp_ping(host, 443, 3).await;
            (code, city, host, ms)
        });

        handles.push(handle);
    }

    let mut results = Vec::new();

    for handle in handles {
        if let Ok((code, city, host, ms)) = handle.await {
            results.push(RegionResult {
                code: code.to_string(),
                city: city.to_string(),
                host: host.to_string(),
                ms,
            });
        }
    }

    results
}

async fn tcp_ping(host: &str, port: u16, attempts: usize) -> f64 {
    let mut samples = Vec::new();

    for _ in 0..attempts {
        let start = Instant::now();
        if let Ok(Ok(_)) = tokio::time::timeout(
            std::time::Duration::from_secs_f64(2.0),
            TcpStream::connect((host, port)),
        )
        .await
        {
            let elapsed = start.elapsed().as_secs_f64() * 1000.0;
            samples.push(elapsed);
        }
    }

    if samples.is_empty() {
        return 0.0;
    }

    round2(samples.iter().cloned().fold(f64::MAX, f64::min))
}

fn round2(v: f64) -> f64 {
    (v * 100.0).round() / 100.0
}
